app/classifier/tasks/survey/index.jsx

```diff
@@ -0,0 +1,182 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import Editor from './editor';
+import Summary from './summary';
+import Chooser from './chooser';
+import Choice from './choice';
+import AnnotationView from './annotation-view';
+
+export default class SurveyTask extends React.Component {
+  static Editor = Editor
+
+  static Summary = Summary
+
+  static AfterSubject = AnnotationView
+
+  static getDefaultTask() {
+    return {
+      type: 'survey',
+      characteristicsOrder: [],
+      characteristics: {},
+      choicesOrder: [],
+      choices: {},
+      questionsOrder: [],
+      questions: {},
+      images: {}
+    };
+  }
+
+  static getTaskText(task) {
+    return `Survey of ${task.choicesOrder.length} choices`;
+  }
+
+  static getDefaultAnnotation() {
+    return { value: [] };
+  }
+
+  static isAnnotationComplete(task, annotation) {
+    const minRequiredAnswers = task.required ? 1 : 0;
+    return annotation.value.length >= minRequiredAnswers && !annotation._choiceInProgress;
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.clearSelection = this.clearSelection.bind(this);
+    this.handleAnnotation = this.handleAnnotation.bind(this);
+    this.handleChoice = this.handleChoice.bind(this);
+    this.handleFilter = this.handleFilter.bind(this);
+    this.handleRemove = this.handleRemove.bind(this);
+
+    this.state = {
+      filters: {},
+      focusedChoice: '',
+      selectedChoiceID: ''
+    };
+  }
+
+  handleFilter(characteristicID, valueID) {
+    const { filters } = this.state;
+
+    setTimeout(() => {
+      if (valueID) {
+        filters[characteristicID] = valueID;
+      } else {
+        delete filters[characteristicID];
+      }
+      this.setState({ filters });
+    });
+  }
+
+  handleChoice(selectedChoiceID) {
+    const { annotation, onChange } = this.props;
+    const newAnnotation = Object.assign({}, annotation, { _choiceInProgress: true });
+
+    this.setState({ selectedChoiceID });
+    onChange(newAnnotation);
+  }
+
+  handleRemove(choiceID) {
+    const { annotation, onChange } = this.props;
+    const selectedChoices = annotation.value.map(item => item.choice);
+    const index = selectedChoices.indexOf(choiceID);
+    annotation.value.splice(index, 1);
+    const newAnnotation = Object.assign({}, annotation);
+    onChange(newAnnotation);
+  }
+
+  clearFilters() {
+    this.setState({ filters: {}});
+  }
+
+  clearSelection() {
+    const { annotation, onChange } = this.props;
+    const { selectedChoiceID } = this.state;
+
+    this.setState({
+      selectedChoiceID: '',
+      focusedChoice: selectedChoiceID
+    });
+    const newAnnotation = Object.assign({}, annotation, { _choiceInProgress: false });
+    onChange(newAnnotation);
+  }
+
+  handleAnnotation(choice, answers, e) {
+    const { filters } = this.state;
+    const { annotation, onChange } = this.props;
+
+    const parsedFilters = JSON.parse(JSON.stringify(filters));
```

Comment:
Copying the filters object by converting it to a string and back might be slow.

Reply:
I'm guessing the alternative is `Object.assign()`? Although, I'm not sure why this wasn't initially used.

Reply:
[`Object.assign()` won't make a new copy of a deeply nested object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Deep_Clone). This code is ok, just slightly puzzling.